swiftide-integrations/src/pgvector/mod.rs

```diff
@@ -0,0 +1,163 @@
+//! This module integrates with the pgvector database, providing functionalities to create and manage vector collections,
+//! store data, and optimize indexing for efficient searches.
+//!
+//! pgvector is utilized in both the `indexing::Pipeline` and `query::Pipeline` modules.
+mod persist;
+mod pgv_table_types;
+use anyhow::Result;
+use derive_builder::Builder;
+use sqlx::PgPool;
+use std::fmt;
+
+use pgv_table_types::{FieldConfig, MetadataConfig, PgDBConnectionPool, VectorConfig};
+
+const DEFAULT_BATCH_SIZE: usize = 50;
+
+/// Represents a Pgvector client with configuration options.
+///
+/// This struct is used to interact with the Pgvector vector database, providing methods to manage vector collections,
+/// store data, and ensure efficient searches. The client can be cloned with low cost as it shares connections.
+#[derive(Builder, Clone)]
+#[builder(setter(into, strip_option), build_fn(error = "anyhow::Error"))]
+pub struct PgVector {
+    /// Database connection pool.
+    #[builder(default = "PgDBConnectionPool::default()")]
+    connection_pool: PgDBConnectionPool,
+
+    /// Table name to store vectors in.
+    #[builder(default = "String::from(\"swiftide_pgv_store\")")]
+    table_name: String,
+
+    /// Default sizes of vectors. Vectors can also be of different
+    /// sizes by specifying the size in the vector configuration.
+    vector_size: Option<i32>,
+
+    /// Batch size for storing nodes.
+    #[builder(default = "Some(DEFAULT_BATCH_SIZE)")]
+    batch_size: Option<usize>,
+
+    /// Field configuration for the Pgvector table, determining the eventual table schema.
+    ///
+    /// Supports multiple field types; see [`FieldConfig`] for details.
+    #[builder(default)]
+    fields: Vec<FieldConfig>,
+}
+
+impl fmt::Debug for PgVector {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Access the connection pool synchronously and determine the status.
+        let connection_status = self.connection_pool.connection_status();
```

Comment:
Does this do a query? Debug is called extensively with `tracing`. Which would mean (I think?) a query on every log / trace statement.